File: src/main/java/com/chessonline/CheckmateHandler.java
package com.chessonline;

import java.util.ArrayList;
import java.util.List;

import com.chessonline.Piece.Position;

public class CheckmateHandler {
    private ChessBoard chessBoard;

    public CheckmateHandler(ChessBoard chessBoard) {
        this.chessBoard = chessBoard;
    }

    public boolean isKingInCheck(King king) {
        if (getThreateningPieces(king).isEmpty()) return false;

        System.out.println("Check for " + king.getColor() + " King");
        return true;
    }

    public List<Piece> getThreateningPieces(King king) {
        List<Piece> threats = new ArrayList<>();
        String kingColor = king.getColor();
        String kingPosition = king.getPosition();
    
        List<Piece> opponentPieces = kingColor.equals("white") ? chessBoard.getBlackPieces() : chessBoard.getWhitePieces();
    
        for (Piece piece : opponentPieces) {
            if (piece.isCaptureMovementValid(kingPosition)) {
                threats.add(piece);
            }
        }
        return threats;
    }

    public boolean isCheckmate(King king) {
        if (!isKingInCheck(king)) return false;
        if (canKingEscape(king)) return false;
        if (canBlockOrCaptureThreats(king)) return false;

        System.out.println("Checkmate for " + king.getColor() + " King");
        return true;
    }

    private boolean canKingEscape(King king) {
        List<String> moves = king.getPossibleMoves();
        for (String move : moves) {
            if (!chessBoard.isDestinationEmpty(move) && chessBoard.isDestinationAlly(move, king.getColor())) continue;
            if (isSquareInCheck(move, king)) continue;
            return true;
        }
        return false;
    }

    private boolean isSquareInCheck(String position, King king) {
        
        int count = 0;
        chessBoard.removePiece(king);

        List<Piece> opponentPieces = king.getColor().equals("white") ? chessBoard.getBlackPieces() : chessBoard.getWhitePieces();
    
        for (Piece piece : opponentPieces) {
            if (piece.isCaptureMovementValid(position)) count++;
        }

        chessBoard.placePiece(king);
        return count > 0;
    }

    private boolean canBlockOrCaptureThreats(King king) {
        List<Piece> threats = getThreateningPieces(king);
        if (threats.size() > 1) return false;
        Piece threat = threats.get(0);
        return handleThreat(threat);
    }

    public boolean handleThreat(Piece threatPiece) {

        List<Piece> allyPieces = (threatPiece.getColor().equals("white")) ? chessBoard.getBlackPieces() : chessBoard.getWhitePieces();
    
        for (Piece ally : allyPieces) {
            if (ally.isCaptureMovementValid(threatPiece.getPosition())) {
                return true;
            }
        }

        if (threatPiece instanceof Knight) return false;
    
        return isBlockPossible(threatPiece);
    }

    private boolean isBlockPossible(Piece threatPiece) {
        King king = (threatPiece.getColor().equals("white")) ? chessBoard.getBlackKing() : chessBoard.getWhiteKing();
        Position kingPos = new Position(king.getPosition());
        Position threatPos = new Position(threatPiece.getPosition());
    
        List<String> path = chessBoard.getPathBetween(kingPos, threatPos);
    
        for (Piece ally : (threatPiece.getColor().equals("white") ? chessBoard.getBlackPieces() : chessBoard.getWhitePieces())) {
            for (String pos : path) {
                if (ally.isMovementValid(pos)) {
                    return true;
                }
            }
        }
    
        return false;
    }
}

File: engine/helpers/checkmate_handler.py
from ..pieces.piece import Position


class CheckmateHandler:
    def __init__(self, chess_board):
        self.chess_board = chess_board

    def is_king_in_check(self, king):
        if not self.get_threatening_pieces(king):
            return False

        print(f"Check for {king.get_color()} King")
        return True

    def get_threatening_pieces(self, king):
        threats = []
        king_color = king.get_color()
        king_position = king.get_position()

        opponent_pieces = (
            self.chess_board.get_black_pieces()
            if king_color == "white"
            else self.chess_board.get_white_pieces()
        )

        for piece in opponent_pieces:
            if piece.is_capture_movement_valid(king_position):
                threats.append(piece)

        return threats

    def is_checkmate(self, king):
        if not self.is_king_in_check(king):
            return False

        if self.can_king_escape(king):
            return False

        if self.can_block_or_capture_threats(king):
            return False

        print(f"Checkmate for {king.get_color()} King")
        return True

    def can_king_escape(self, king):
        moves = king.get_possible_moves()

        for move in moves:
            if (
                not self.chess_board.is_destination_empty(move)
                and self.chess_board.is_destination_ally(move, king.get_color())
            ):
                continue

            if self.is_square_in_check(move, king):
                continue

            return True

        return False

    def is_square_in_check(self, position, king):
        count = 0

        self.chess_board.remove_piece(king)

        opponent_pieces = (
            self.chess_board.get_black_pieces()
            if king.get_color() == "white"
            else self.chess_board.get_white_pieces()
        )

        for piece in opponent_pieces:
            if piece.is_capture_movement_valid(position):
                count += 1

        self.chess_board.place_piece(king)

        return count > 0

    def can_block_or_capture_threats(self, king):
        threats = self.get_threatening_pieces(king)

        if len(threats) > 1:
            return False

        threat = threats[0]
        return self.handle_threat(threat)

    def handle_threat(self, threat_piece):
        ally_pieces = (
            self.chess_board.get_black_pieces()
            if threat_piece.get_color() == "white"
            else self.chess_board.get_white_pieces()
        )

        # can capture the threatening piece
        for ally in ally_pieces:
            if ally.is_capture_movement_valid(threat_piece.get_position()):
                return True

        # knight cannot be blocked (knights jump over pieces)
        # TODO: replace with isinstance(threat_piece, Knight) once Knight is implemented
        if threat_piece.get_name() == "knight":
            return False

        return self.is_block_possible(threat_piece)

    def is_block_possible(self, threat_piece):
        king = (
            self.chess_board.get_black_king()
            if threat_piece.get_color() == "white"
            else self.chess_board.get_white_king()
        )

        king_pos = Position(king.get_position())
        threat_pos = Position(threat_piece.get_position())

        path = self.chess_board.get_path_between(king_pos, threat_pos)

        ally_pieces = (
            self.chess_board.get_black_pieces()
            if threat_piece.get_color() == "white"
            else self.chess_board.get_white_pieces()
        )

        for ally in ally_pieces:
            for pos in path:
                if ally.is_movement_valid(pos):
                    return True

        return False